File: results/2026-04-08_113116/tasks/11-semantic-version-bumper/default-opus/generated-code/version_bumper.py
"""Semantic version bumper.

Parses version files, classifies conventional commits, bumps the version,
generates a changelog entry, and outputs the new version.
"""

import re
import json
import os
import sys
from datetime import date
from pathlib import Path


# --- Version parsing ---

def parse_version(version_str: str) -> tuple[int, int, int]:
    """Parse a semver string like '1.2.3' or 'v1.2.3' into (major, minor, patch)."""
    match = re.match(r"^v?(\d+)\.(\d+)\.(\d+)$", version_str.strip())
    if not match:
        raise ValueError(f"Invalid semantic version: '{version_str}'")
    return int(match.group(1)), int(match.group(2)), int(match.group(3))


# --- Commit classification ---

# Conventional commit pattern: type(optional-scope)optional-!: description
_COMMIT_RE = re.compile(r"^(?P<type>\w+)(?:\([^)]*\))?(?P<bang>!)?:\s")

# Only these types trigger a version bump
_BUMP_TYPES = {"feat": "minor", "fix": "patch", "perf": "patch"}


def classify_commits(commits: list[str]) -> dict[str, list[str]]:
    """Classify conventional commits into {major, minor, patch} buckets.

    A commit is 'major' if it has a '!' after the type/scope or contains
    'BREAKING CHANGE:' in the body. 'feat' maps to minor, 'fix'/'perf' to patch.
    Non-bumping types (docs, chore, style, etc.) are ignored.
    """
    result: dict[str, list[str]] = {}
    for commit in commits:
        first_line = commit.split("\n")[0]
        match = _COMMIT_RE.match(first_line)
        if not match:
            continue

        # Check for breaking change indicators
        if match.group("bang") or "BREAKING CHANGE:" in commit:
            result.setdefault("major", []).append(commit)
            continue

        bump = _BUMP_TYPES.get(match.group("type"))
        if bump:
            result.setdefault(bump, []).append(commit)

    return result


def determine_bump(classified: dict[str, list[str]]) -> str | None:
    """Return the highest-priority bump level, or None if no bump needed."""
    for level in ("major", "minor", "patch"):
        if level in classified:
            return level
    return None


# --- Version bumping ---

def bump_version(version: tuple[int, int, int], level: str | None) -> tuple[int, int, int]:
    """Apply a bump level to a version tuple. Returns unchanged version if level is None."""
    major, minor, patch = version
    if level == "major":
        return (major + 1, 0, 0)
    elif level == "minor":
        return (major, minor + 1, 0)
    elif level == "patch":
        return (major, minor, patch + 1)
    return version


def format_version(version: tuple[int, int, int]) -> str:
    """Format a version tuple as 'X.Y.Z'."""
    return f"{version[0]}.{version[1]}.{version[2]}"


# --- Version file I/O ---

def read_version_file(path: Path) -> str:
    """Read a version string from a plain VERSION file or package.json."""
    path = Path(path)
    if not path.exists():
        raise FileNotFoundError(f"Version file not found: {path}")

    if path.name == "package.json":
        data = json.loads(path.read_text())
        if "version" not in data:
            raise ValueError(f"No 'version' field in {path}")
        return data["version"]

    # Plain text version file
    return path.read_text().strip()


def write_version_file(path: Path, new_version: str) -> None:
    """Write a version string to a plain VERSION file or package.json."""
    path = Path(path)
    if path.name == "package.json":
        data = json.loads(path.read_text())
        data["version"] = new_version
        path.write_text(json.dumps(data, indent=2) + "\n")
    else:
        path.write_text(new_version + "\n")


# --- Changelog generation ---

# Map bump categories to human-readable changelog section names
_SECTION_NAMES = {
    "major": "BREAKING CHANGES",
    "minor": "Features",
    "patch": "Bug Fixes",
}


def _extract_description(commit: str) -> str:
    """Pull the short description out of a conventional commit message."""
    first_line = commit.split("\n")[0]
    # Strip 'type(scope)!: ' prefix to get just the description
    match = _COMMIT_RE.match(first_line)
    if match:
        return first_line[match.end():]
    return first_line


def generate_changelog(version: str, classified: dict[str, list[str]]) -> str:
    """Produce a markdown changelog entry for the given version and classified commits."""
    today = date.today().isoformat()
    lines = [f"## {version} ({today})", ""]

    # Emit sections in priority order so breaking changes appear first
    for level in ("major", "minor", "patch"):
        commits = classified.get(level, [])
        if not commits:
            continue
        lines.append(f"### {_SECTION_NAMES[level]}")
        lines.append("")
        for c in commits:
            lines.append(f"- {_extract_description(c)}")
        lines.append("")

    return "\n".join(lines)


# --- Fixture loading ---

def load_commits_from_file(path: Path) -> list[str]:
    """Load commit messages from a fixture file.

    Commits are separated by '---' lines, allowing multi-line commit
    bodies (e.g., with BREAKING CHANGE: footers) within a single entry.
    """
    path = Path(path)
    text = path.read_text()

    chunks = re.split(r"\n---\n", text.strip())
    return [chunk.strip() for chunk in chunks if chunk.strip()]


# --- CLI entry point ---

def main() -> None:
    """CLI: version_bumper.py <version-file> <commit-log-file>

    Reads the current version, classifies commits, bumps the version,
    writes the updated file, prints changelog and new version to stdout.
    """
    if len(sys.argv) < 3:
        print(
            "Usage: version_bumper.py <version-file> <commit-log-file>",
            file=sys.stderr,
        )
        sys.exit(1)

    version_file = Path(sys.argv[1])
    commit_file = Path(sys.argv[2])

    # Read current version
    raw_version = read_version_file(version_file)
    current = parse_version(raw_version)

    # Classify commits and determine bump
    commits = load_commits_from_file(commit_file)
    classified = classify_commits(commits)
    level = determine_bump(classified)

    if level is None:
        print("No version-bumping commits found. Version unchanged.")
        print(format_version(current))
        sys.exit(0)

    # Bump and write
    new = bump_version(current, level)
    new_str = format_version(new)
    write_version_file(version_file, new_str)

    # Generate and print changelog
    changelog = generate_changelog(new_str, classified)
    print(changelog)
    print(f"Version bumped: {format_version(current)} -> {new_str} ({level})")


if __name__ == "__main__":
    main()

File: results/2026-04-09_152435/tasks/11-semantic-version-bumper/typescript-bun-sonnet/generated-code/src/version-bumper.ts
/**
 * Semantic Version Bumper
 *
 * Implements conventional-commits-based semantic versioning:
 * - "fix:" prefix  → patch bump (1.0.0 → 1.0.1)
 * - "feat:" prefix → minor bump (1.0.0 → 1.1.0)
 * - "type!:" or "BREAKING CHANGE:" → major bump (1.0.0 → 2.0.0)
 *
 * Usage (CLI):
 *   bun run src/version-bumper.ts fixtures/test-case-1.json
 *
 * The test-case JSON format:
 * {
 *   "currentVersion": "1.0.0",
 *   "commits": [ { "hash": "abc1234", "message": "fix: ...", "author": "...", "date": "..." } ]
 * }
 */

// ─── Types ────────────────────────────────────────────────────────────────────

export interface SemanticVersion {
  major: number;
  minor: number;
  patch: number;
}

export interface Commit {
  hash: string;
  message: string;
  author: string;
  date: string;
}

export interface TestCase {
  currentVersion: string;
  commits: Commit[];
  expectedNewVersion?: string; // Optional: for assertion in CI
}

export type BumpType = "major" | "minor" | "patch" | "none";

// ─── parseVersion ─────────────────────────────────────────────────────────────

/**
 * Parse a "X.Y.Z" semver string into a SemanticVersion object.
 * Throws a clear error for malformed strings.
 */
export function parseVersion(versionString: string): SemanticVersion {
  const match = versionString.match(/^(\d+)\.(\d+)\.(\d+)$/);
  if (!match) {
    throw new Error(`Invalid version string: "${versionString}" (expected X.Y.Z format)`);
  }
  return {
    major: parseInt(match[1], 10),
    minor: parseInt(match[2], 10),
    patch: parseInt(match[3], 10),
  };
}

// ─── formatVersion ────────────────────────────────────────────────────────────

/**
 * Format a SemanticVersion object into "X.Y.Z" string.
 */
export function formatVersion(version: SemanticVersion): string {
  return `${version.major}.${version.minor}.${version.patch}`;
}

// ─── determineBumpType ────────────────────────────────────────────────────────

/**
 * Analyze an array of conventional commits to determine the highest bump type.
 *
 * Conventional commit rules:
 * - "fix:"   or "fix(scope):"   → patch
 * - "feat:"  or "feat(scope):"  → minor
 * - "type!:" or body contains "BREAKING CHANGE:" → major (immediate return)
 *
 * Priority: major > minor > patch > none
 */
export function determineBumpType(commits: Commit[]): BumpType {
  let bumpType: BumpType = "none";

  for (const commit of commits) {
    const msg = commit.message;

    // Breaking change: type followed by "!" OR "BREAKING CHANGE:" in body
    // Check this first since it short-circuits (major always wins)
    if (/^[a-z]+(\([^)]+\))?!:/.test(msg) || msg.includes("BREAKING CHANGE:")) {
      return "major";
    }

    // feat (with or without scope) → minor
    if (/^feat(\([^)]+\))?:/.test(msg)) {
      bumpType = "minor";
      continue;
    }

    // fix (with or without scope) → patch (only if we haven't found minor yet)
    if (/^fix(\([^)]+\))?:/.test(msg) && bumpType === "none") {
      bumpType = "patch";
    }
  }

  return bumpType;
}

// ─── bumpVersion ──────────────────────────────────────────────────────────────

/**
 * Apply a bump type to a version, following semver rules:
 * - major: increment major, reset minor+patch to 0
 * - minor: increment minor, reset patch to 0
 * - patch: increment patch
 * - none: return version unchanged
 */
export function bumpVersion(version: SemanticVersion, bumpType: BumpType): SemanticVersion {
  switch (bumpType) {
    case "major":
      return { major: version.major + 1, minor: 0, patch: 0 };
    case "minor":
      return { major: version.major, minor: version.minor + 1, patch: 0 };
    case "patch":
      return { major: version.major, minor: version.minor, patch: version.patch + 1 };
    case "none":
      return { ...version };
  }
}

// ─── generateChangelog ────────────────────────────────────────────────────────

/**
 * Generate a markdown changelog entry from commits for the given new version.
 * Groups commits into: Breaking Changes, Features, Bug Fixes, Other.
 */
export function generateChangelog(commits: Commit[], newVersion: string): string {
  const date = new Date().toISOString().split("T")[0];
  const lines: string[] = [`## [${newVersion}] - ${date}`, ""];

  const breaking: string[] = [];
  const features: string[] = [];
  const fixes: string[] = [];
  const other: string[] = [];

  for (const commit of commits) {
    const msg = commit.message;
    // Use first 7 chars of hash for brevity (like git log --abbrev)
    const shortHash = commit.hash.slice(0, 7);
    const entry = `- ${msg.split("\n")[0]} (${shortHash})`;

    if (/^[a-z]+(\([^)]+\))?!:/.test(msg) || msg.includes("BREAKING CHANGE:")) {
      breaking.push(entry);
    } else if (/^feat(\([^)]+\))?:/.test(msg)) {
      features.push(entry);
    } else if (/^fix(\([^)]+\))?:/.test(msg)) {
      fixes.push(entry);
    } else {
      other.push(entry);
    }
  }

  if (breaking.length > 0) {
    lines.push("### Breaking Changes");
    lines.push(...breaking);
    lines.push("");
  }
  if (features.length > 0) {
    lines.push("### Features");
    lines.push(...features);
    lines.push("");
  }
  if (fixes.length > 0) {
    lines.push("### Bug Fixes");
    lines.push(...fixes);
    lines.push("");
  }
  if (other.length > 0) {
    lines.push("### Other");
    lines.push(...other);
    lines.push("");
  }

  return lines.join("\n");
}

// ─── CLI Entry Point ──────────────────────────────────────────────────────────

/**
 * When run directly with `bun run src/version-bumper.ts <test-case-file>`:
 * - Reads the test case JSON file
 * - Determines the new version from commits
 * - Prints: NEW_VERSION: X.Y.Z
 * - Prints the changelog entry
 * - Exits with non-zero if expected version doesn't match
 */
if (import.meta.main) {
  const testCaseFile = Bun.argv[2];

  if (!testCaseFile) {
    console.error("Usage: bun run src/version-bumper.ts <test-case-file.json>");
    console.error("  The JSON file must have: { currentVersion, commits, expectedNewVersion? }");
    process.exit(1);
  }

  const file = Bun.file(testCaseFile);
  const exists = await file.exists();
  if (!exists) {
    console.error(`Error: File not found: ${testCaseFile}`);
    process.exit(1);
  }

  const testCase: TestCase = await file.json();

  // Validate input
  if (!testCase.currentVersion) {
    console.error("Error: testCase.currentVersion is required");
    process.exit(1);
  }
  if (!Array.isArray(testCase.commits)) {
    console.error("Error: testCase.commits must be an array");
    process.exit(1);
  }

  const version = parseVersion(testCase.currentVersion);
  const bumpType = determineBumpType(testCase.commits);
  const newVersion = bumpVersion(version, bumpType);
  const newVersionStr = formatVersion(newVersion);
  const changelog = generateChangelog(testCase.commits, newVersionStr);

  // Structured output that the workflow can grep for
  console.log(`NEW_VERSION: ${newVersionStr}`);
  console.log(`BUMP_TYPE: ${bumpType}`);
  console.log("");
  console.log("CHANGELOG:");
  console.log(changelog);

  // Optional: assert expected version matches if provided in fixture
  if (testCase.expectedNewVersion !== undefined) {
    if (newVersionStr !== testCase.expectedNewVersion) {
      console.error(
        `ERROR: Version mismatch! Expected ${testCase.expectedNewVersion}, got ${newVersionStr}`
      );
      process.exit(1);
    }
    console.log(`ASSERTION PASSED: ${testCase.expectedNewVersion}`);
  }
}
